### PR TITLE
feature: search/filter functionality in marketplace

### PR DESCRIPTION
**Pull request by ebarthur**

### CODE REVIEW

**Comment by ebarthur** [comment]
Search/filter changes now updates the request URL @blackmann

File: client/app/routes/market.tsx [hunk 1]
```diff
@@ -20,12 +26,32 @@ export const loader = async ({ request }: LoaderFunctionArgs) => {
...
+		ignoreQueryPrefix: true,
+	});
+
+	const whereClause: any = {
+		status: "available",
+	};
+
+	if (q) {
+		whereClause.OR = [
+			{ name: { contains: q.toString(), lte: "insensitive" } },
```
Comment (blackmann) [inline client/app/routes/market.tsx]: What does `lte: "insensitive` do?

File: client/app/routes/market.tsx [hunk 2]
```diff
@@ -20,12 +26,32 @@ export const loader = async ({ request }: LoaderFunctionArgs) => {
...
+			{ description: { contains: q.toString(), lte: "insensitive" } },
+		];
+	}
+
+	if (category && category !== "all") {
```
Comment (blackmann) [inline client/app/routes/market.tsx]:
```suggestion
	if (category !== "all") {
```

should be enough

File: client/app/routes/market.tsx [hunk 3]
```diff
@@ -49,18 +75,43 @@ export default function Market() {
...
+
+	const [searchQuery, setSearchQuery] = useState<string>("");
+	const [selectedCategory, setSelectedCategory] = useState<string>("all");
+
+	const handleSearchInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+		const value = e.target.value;
+		setSearchQuery(value);
+		navigate(
+			`/market?q=${encodeURIComponent(value)}&category=${encodeURIComponent(selectedCategory)}`, // Encode category name
+		);
```
Comment (blackmann) [inline client/app/routes/market.tsx]: Move this to an effect (ie `useEffect`)

And also when you refresh the page and the params are in the URL, the search field and the category should reflect those values.

File: client/app/routes/market.tsx [hunk 4]
```diff
@@ -49,18 +75,43 @@ export default function Market() {
...
+						type="search"
+						placeholder="Search product"
+						value={searchQuery}
+						onChange={handleSearchInputChange}
+					/>
+					<Select value={selectedCategory} onChange={handleCategoryChange}>
 						<option value="all">All Categories</option>
 						{categories.map((category) => (
-							<option key={category.id} value={category.id}>
+							<option key={category.id} value={category.title}>
```
Comment (blackmann) [inline client/app/routes/market.tsx]: Let's maintain the `id`

The prisma filter would be `categoryId`

**Review by blackmann — CHANGES_REQUESTED** [review]
Good start.

Please address the comments, lesgoo

File: client/app/routes/market.tsx (hunk 1 above)
Comment (ebarthur) [reply client/app/routes/market.tsx]: Case-sensitive matching in Prisma. "Mode: "insensitive"" was throwing an error. It's not even necessary; It works without it anyway. I'll remove that

File: client/app/routes/market.tsx (hunk 2 above)
Comment (ebarthur) [reply client/app/routes/market.tsx]: Alright. Thanks.

**Comment by ebarthur** [comment]
All comments have been resolved. It works fine with debounce similar to the code in library.tsx. Thank you

File: client/app/routes/market.tsx [hunk 5]
```diff
@@ -20,12 +27,32 @@ export const loader = async ({ request }: LoaderFunctionArgs) => {
...
+
+	if (q) {
+		whereClause.OR = [
+			{ name: { contains: q.toString() } },
+			{ description: { contains: q.toString() } },
+		];
+	}
+
+	if (category !== "all") {
+		whereClause.category = { title: category?.toString() };
```
Comment (blackmann) [inline client/app/routes/market.tsx]: Category is id in the params

File: client/app/routes/market.tsx [hunk 6]
```diff
@@ -49,15 +76,74 @@ export default function Market() {
...
+
+	const searchParams = new URLSearchParams(location.search);
+	const initialQuery = searchParams.get("q") || "";
+	const initialCategory = searchParams.get("category") || "all";
+
+	const [searchQuery, setSearchQuery] = useState<string>(initialQuery);
+	const [selectedCategory, setSelectedCategory] =
+		useState<string>(initialCategory);
+
+	useEffect(() => {
```
Comment (blackmann) [inline client/app/routes/market.tsx]: For consistency with the rest of the codebase, let's prefix React hooks with `React.`

**Review by blackmann — CHANGES_REQUESTED** [review]
Works well except the category filter.

+ some code suggestion. then I merge

File: client/app/routes/market.tsx (hunk 6 above)
Comment (ebarthur) [reply client/app/routes/market.tsx]: Fixed

**Pull request merged**